src/Azure/Orleans.ScheduledJobs.AzureStorage/AzureStorageJobShardManager.cs

```diff
@@ -0,0 +1,396 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure;
+using Azure.Storage.Blobs;
+using Azure.Storage.Blobs.Models;
+using Azure.Storage.Blobs.Specialized;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Orleans.Hosting;
+using Orleans.Runtime;
+
+namespace Orleans.ScheduledJobs.AzureStorage;
+
+public sealed partial class AzureStorageJobShardManager : JobShardManager
+{
+    private readonly BlobServiceClient _blobServiceClient;
+    private readonly string _containerName;
+    private readonly string _blobPrefix;
+    private BlobContainerClient _client = null!;
+    private readonly IClusterMembershipService _clusterMembership;
+    private readonly ConcurrentDictionary<string, AzureStorageJobShard> _jobShardCache = new();
+    private readonly ILogger<AzureStorageJobShardManager> _logger;
+
+    public AzureStorageJobShardManager(
+        BlobServiceClient client,
+        string containerName,
+        string blobPrefix,
+        IClusterMembershipService clusterMembership,
+        ILogger<AzureStorageJobShardManager> logger)
+    {
+        _blobServiceClient = client;
+        _containerName = containerName;
+        _blobPrefix = blobPrefix;
+        _clusterMembership = clusterMembership;
+        _logger = logger;
+    }
+
+    public AzureStorageJobShardManager(
+        ILocalSiloDetails localSiloDetails,
+        IOptions<AzureStorageJobShardOptions> options,
+        IClusterMembershipService clusterMembership,
+        ILogger<AzureStorageJobShardManager> logger)
+        : this(options.Value.BlobServiceClient, options.Value.ContainerName, localSiloDetails.ClusterId, clusterMembership, logger)
+    {
+    }
+
+    public override async Task<List<IJobShard>> AssignJobShardsAsync(SiloAddress siloAddress, DateTimeOffset maxShardStartTime, CancellationToken cancellationToken = default)
```

Comment:
This should be minimally greedy, like `LeaseBasedQueueBalancer` is. i.e, take up to 1 more shard than our fair share based on current membership.